### PR TITLE
[Pytorch][Build][Test] Fix PT1.9.1 training vuls

### PR DESCRIPTION
**Pull request by shantanutrip**
*GitHub Issue #, if available:*

Note: If merging this PR should also close the associated Issue, please also add that Issue # to the Linked Issues section on the right. 

### Description
For fixing issues with PT1.9.1 images

### Tests run
**NOTE: If you are creating a PR for a new framework version, please ensure success of the standard, rc, and efa sagemaker remote tests by updating the dlc_developer_config.toml file:**

- [ ] Revision A: `sagemaker_remote_tests = "standard"`
- [ ] Revision B: `sagemaker_remote_tests = "rc"`
- [ ] Revision C: `sagemaker_remote_tests = "efa"`

**Additionally, please run the sagemaker local tests in at least one revision:**
- [ ] `sagemaker_local_tests = true`

## Label Checklist
- [ ] I have added the project label for this PR (*<project_name>* or "Improvement")

## PR Checklist
- [ ] I've prepended PR tag with frameworks/job this applies to : [mxnet, tensorflow, pytorch] | [ei/neuron/graviton] | [build] | [test] | [benchmark] | [ec2, ecs, eks, sagemaker]
- [ ] If the PR changes affects SM test, I've modified dlc_developer_config.toml in my PR branch by setting sagemaker_tests = true and efa_tests = true
- [ ] If this PR changes existing code, the change fully backward compatible with pre-existing code. (Non backward-compatible changes need special approval.)
- [ ] (If applicable) I've documented below the DLC image/dockerfile this relates to
- [ ] (If applicable) I've documented below the tests I've run on the DLC image
- [ ] (If applicable) I've reviewed the licenses of updated and new binaries and their dependencies to make sure all licenses are on the Apache Software Foundation Third Party License Policy Category A or Category B license list.  See [https://www.apache.org/legal/resolved.html](https://www.apache.org/legal/resolved.html).
- [ ] (If applicable) I've scanned the updated and new binaries to make sure they do not have vulnerabilities associated with them.

## Pytest Marker Checklist
- [ ] (If applicable) I have added the marker `@pytest.mark.model("<model-type>")` to the new tests which I have added, to specify the Deep Learning model that is used in the test (use `"N/A"` if the test doesn't use a model)
- [ ] (If applicable) I have added the marker `@pytest.mark.integration("<feature-being-tested>")` to the new tests which I have added, to specify the feature that will be tested
- [ ] (If applicable) I have added the marker `@pytest.mark.multinode(<integer-num-nodes>)` to the new tests which I have added, to specify the number of nodes used on a multi-node test
- [ ] (If applicable) I have added the marker `@pytest.mark.processor(<"cpu"/"gpu"/"eia"/"neuron">)` to the new tests which I have added, if a test is specifically applicable to only one processor type

#### EIA/NEURON/GRAVITON Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `ei_mode = true`, `neuron_mode = true` or `graviton_mode = true`

#### Benchmark Testing Checklist
* When creating a PR:
- [ ] I've modified `dlc_developer_config.toml` in my PR branch by setting `benchmark_mode = true`

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license. I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py [hunk 1]
```diff
@@ -23,15 +24,23 @@
...
 
-# set default instance type to p3.16
-if "SAGEMAKER_INSTANCE_TYPE" not in os.environ:
-    os.environ['SAGEMAKER_INSTANCE_TYPE'] = 'ml.p3.16xlarge'
+TORCH_VERSION = torch.__version__
 
-dist.init_process_group(backend='smddp')
+if Version(TORCH_VERSION) < Version("1.10"):
+    from smdistributed.dataparallel.torch.parallel.distributed import DistributedDataParallel as DDP
+    import smdistributed.dataparallel.torch.distributed as dist
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py:32]: do we need a corresponding import for `import smdistributed.dataparallel.torch.torch_smddp`

File: test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py [hunk 2]
```diff
@@ -23,15 +24,23 @@
...
+    dist.init_process_group()
+else:
+    from torch.nn.parallel import DistributedDataParallel as DDP
+    import torch.distributed as dist
+    import smdistributed.dataparallel.torch.torch_smddp
+
+    # set default instance type to p3.16
```
Comment (arjkesh) [inline test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py:39]: why ml.p3 by default?

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py [hunk 3]
```diff
@@ -170,6 +170,9 @@ def test_smmodelparallel_mnist_multigpu_singlenode(ecr_image, instance_type, sag
     """
     Tests pt gpt2 command via script mode
     """
+    framework, framework_version = get_framework_and_version_from_tag(ecr_image)
+    if framework == "pytorch" and Version(framework_version) == Version("1.9.1"):
+        pytest.skip("Skipping the test for PT1.9")
```
Comment (saimidu) [inline test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py:175]: Please do not use `Version(variable_version) == Version("x.y.z")` in code. It is highly restrictive, and will cause the condition to work incorrectly the moment a new patch version is used.

Please always use `Version(framework_version) in SpecifierSet("==1.9.*")` instead.

File: test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py (hunk 1 above)
Comment (saimidu) [reply test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py:32]: +1

File: test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py (hunk 2 above)
Comment (saimidu) [reply test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py:39]: +1

File: test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py (hunk 1 above)
Comment (shantanutrip) [reply test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py:32]: Changes made during PT1.10 release turned the following script incompatible with PT1.9 version. So, the changes that I have made in this file are to make the script backward compatible. To do that, based on the DataParallel team's suggestion, I looked into the script that existed when PT1.9 was released (https://github.com/shantanutrip/deep-learning-containers/blob/b87590af021a4c3913b3669fa5f9fc10fbebe4e8/test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py) and found its diff with the current file  on the master branch (https://github.com/aws/deep-learning-containers/blob/master/test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py) and made the changes accordingly. 

For the old file, I do not see any `import smdistributed.dataparallel.torch.torch_smddp` line or any other import corresponding to it, so I did not add it.

File: test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py (hunk 3 above)
Comment (shantanutrip) [reply test/sagemaker_tests/pytorch/training/integration/sagemaker/test_distributed_operations.py:175]: I agree, the `SpecifierSet("==1.9.*")` solution is a good practice. Thanks, will make the change.

File: test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py (hunk 2 above)
Comment (shantanutrip) [reply test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py:39]: AS stated here: https://github.com/aws/deep-learning-containers/pull/1948#discussion_r885216598 , I just merged the 2 files logically. I think the SMDataParallel team would have a better answer to it.

File: test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py [hunk 4]
```diff
@@ -23,15 +24,23 @@
...
-import torch.distributed as dist
-import smdistributed.dataparallel.torch.torch_smddp
 
-# set default instance type to p3.16
-if "SAGEMAKER_INSTANCE_TYPE" not in os.environ:
-    os.environ['SAGEMAKER_INSTANCE_TYPE'] = 'ml.p3.16xlarge'
+TORCH_VERSION = torch.__version__
 
-dist.init_process_group(backend='smddp')
+if Version(TORCH_VERSION) < Version("1.10"):
```
Comment (indhub) [inline test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py:30]: We should consider ways to avoid branching based on version in test files. These ifdefs in the test file could gets more complicated as more versions with breaking changes are released.

File: test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py (hunk 1 above)
Comment (indhub) [reply test/sagemaker_tests/pytorch/training/resources/mnist/smdataparallel_mnist.py:32]: Older SMDDP versions don't need an equivalent import

**Review by saimidu — DISMISSED** [review]
LGTM. Please revert temp changes.
